src/Microsoft.Dnx.Watcher.Core/ProjectWatcher.cs

```diff
@@ -0,0 +1,286 @@
+                        return project;
+                    }
+
+                    _logger.LogError($"Error(s) reading project file '{_fullProjectFilePath}': ");
+                    _logger.LogError(errors);
+                    _logger.LogInformation("Fix the error to continue.");
+
+                    await WatchForFileChangeAsync(fileWatcher, cancellationToken);
+
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        return null;
+                    }
+
+                    _logger.LogInformation($"File changed: {_fullProjectFilePath}");
+                }
+            }
+        }
+
+        private void AddProjectAndDependeciesToWatcher(Project project, IFileWatcher fileWatcher)
+        {
+            // TODO: we need the framework to run on
+            var libManager = _projectGraphProvider.GetProjectGraph(
+                project,
+                new FrameworkName(FrameworkNames.LongNames.Dnx, new Version(4, 5, 1)));
+
+            var projectsToWatch = libManager.GetLibraryDescriptions().OfType<ProjectDescription>()
+                .Select(lib => lib.Project);
+
+            foreach (var proj in projectsToWatch)
+            {
+                var filesToWatch =
+                    proj.Files.SourceFiles.Concat(
+                    proj.Files.ResourceFiles.Values.Concat(
+                    proj.Files.PreprocessSourceFiles.Concat(
+                    proj.Files.SharedFiles)));
+
+                foreach (var file in filesToWatch)
+                {
+                    fileWatcher.WatchDirectory(
+                        Path.GetDirectoryName(file), 
+                        Path.GetExtension(file));
+                }
+                
+                fileWatcher.WatchProject(proj.ProjectFilePath);
+            }
+
+            fileWatcher.WatchFile(project.ProjectFilePath);
+            fileWatcher.WatchProject(project.ProjectFilePath);
+        }
+
+        // Same as TryGetProject but it doesn't throw
+        private bool TryGetProject(string projectFile, out Project project, out string errorMessage)
+        {
+            try
+            {
+                ICollection<DiagnosticMessage> errors = new List<DiagnosticMessage>();
+                if (!Project.TryGetProject(_fullProjectFilePath, out project, errors))
+                {
+                    errorMessage = string.Join(Environment.NewLine, errors.Select(e => e.ToString()));
+                }
+                else
+                {
+                    errorMessage = null;
+                    return true;
+                }
+            }
+            catch (Exception ex)
+            {
+                errorMessage = ex.Message;
+            }
+
+            project = null;
+            return false;
+        }
+
+        private async Task<string> WatchForFileChangeAsync(IFileWatcher fileWatcher, CancellationToken cancellationToken)
+        {
+            string changedFile = null;
```

Comment:
Use a TaskCompletionSource and cancellationToken.Register instead of this logic:

```C#
var tcs = new TaskCompletionSource<string>();
fileWatcher.OnChanged += path => tcs.TrySetResult(path);
cancellationToken.Register(() => tcs.TrySetCancelled());
return tcs.Task;
```